src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverter.ReadAhead.cs

```diff
@@ -0,0 +1,72 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.Text.Json.Serialization
+{
+    /// <summary>
+    /// Converts an object or value to or from JSON.
+    /// </summary>
+    public abstract partial class JsonConverter
+    {
+        // AggressiveInlining used since this method is on a hot path and short. The optionally called
+        // method DoSingleValueReadWithReadAhead is not inlined.
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static bool SingleValueReadWithReadAhead(ClassType classType, ref Utf8JsonReader reader, ref ReadStack state)
+        {
+            bool readAhead = (state.ReadAhead && (classType & (ClassType.Value | ClassType.NewValue)) != 0);
```

Comment:
Yes

Reply:
nit: Every time I read these conditions, I have to remind myself what it means (especially since I think it is negation of what it actually means) :p
A comment or Debug.Assert showcasing the "simple" condition would help readability.

We can do a pass for that outside this PR.